### PR TITLE
[WIP] Multilingual

### PR DESCRIPTION
**Pull request by creatorscan**

### CODE REVIEW

**Comment by sw005320** [comment]
Can you first pass the travis-ci check?

**Comment by creatorscan** [comment]
Yeah, I will do that and let you know.

**Comment by sw005320** [comment]
See https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md to deal with the travis-ci check.

**Comment by creatorscan** [comment]
Thanks for the document Shinji, it really helped to spot the style fixes..

**Comment by creatorscan** [comment]
@sw005320 , Conflicts are resolved.. please merge it when possible.. Thanks

File: src/bin/asr_train.py [hunk 1]
```diff
@@ -37,6 +37,17 @@ def main():
                         help='Output directory for debugging')
     parser.add_argument('--resume', '-r', default='', nargs='?',
                         help='Resume the training from snapshot')
+    parser.add_argument('--modify_output', '-m', default='false', type=bool,
+                        help='Wehether to modify output layer to new language')
+    parser.add_argument('--adapt', '-a', type=bool, default=False,
+                        help='retrain/adapt from the exisitng best model')
+    parser.add_argument('--adaptLayerNames', '-l', type=str, default="AttCtcOut",
```
Comment (sw005320) [inline src/bin/asr_train.py]: `--adapt-layer-names` is more consistent with other option name conventions.

File: src/bin/asr_train.py [hunk 2]
```diff
@@ -37,6 +37,17 @@ def main():
                         help='Output directory for debugging')
     parser.add_argument('--resume', '-r', default='', nargs='?',
                         help='Resume the training from snapshot')
```
Comment (sw005320) [inline src/bin/asr_train.py]: better to add a comment when adaptation related arguments are started, e.g.,
```
    parser.add_argument('--resume', '-r', default='', nargs='?',
                        help='Resume the training from snapshot')
    # adaptation
    parser.add_argument('--modify_output', '-m', default='false', type=bool,
                        help='Wehether to modify output layer to new language')
```

File: src/bin/asr_train.py [hunk 3]
```diff
@@ -37,6 +37,17 @@ def main():
...
+    parser.add_argument('--modify_output', '-m', default='false', type=bool,
```
Comment (sw005320) [inline src/bin/asr_train.py]: `--modify_output` -> `--modify-output`

File: src/bin/asr_train.py [hunk 4]
```diff
@@ -37,6 +37,17 @@ def main():
...
+                        help='Wehether to modify output layer to new language')
+    parser.add_argument('--adapt', '-a', type=bool, default=False,
+                        help='retrain/adapt from the exisitng best model')
+    parser.add_argument('--adaptLayerNames', '-l', type=str, default="AttCtcOut",
+                        choices=['AttOut', 'Out', 'AttCtc', 'AttCtcOut', 'CtcOut'],
+                        help='only retrain these layers from the exisitng best model')
+    parser.add_argument('--freeze', '-f', type=bool, default=False,
+                        help='freeze layers from the exisitng best model')
+    parser.add_argument('--args.noencs_freeze', '-ne', type=int, default=8,
```
Comment (sw005320) [inline src/bin/asr_train.py]: `--args.noencs_freeze` -> `--noencs-freeze`

File: src/bin/asr_train.py [hunk 5]
```diff
@@ -105,6 +116,9 @@ def main():
                         help='Apply label smoothing with a specified distribution type')
     parser.add_argument('--lsm-weight', default=0.0, type=float,
                         help='Label smoothing weight')
+    # Scheduled sampling flags
+    parser.add_argument('--tf-ratio', default=1.0, type=float,
```
Comment (sw005320) [inline src/bin/asr_train.py]: what does `tf` mean? Better to add more explanation in help

File: src/bin/asr_train.py [hunk 6]
```diff
@@ -117,8 +131,16 @@ def main():
                         help='Batch size is reduced if the output sequence length > ML')
     # optimization related
     parser.add_argument('--opt', default='adadelta', type=str,
-                        choices=['adadelta', 'adam'],
+                        choices=['adadelta', 'adam', 'sgd'],
                         help='Optimizer')
+    parser.add_argument('--lr', default=1e-3, type=float,
+                        help='Learning rate constant for sgd optimizer')
+    parser.add_argument('--lr_decay', default=1e-1, type=float,
```
Comment (sw005320) [inline src/bin/asr_train.py]: `--lr_decay` -> `--lr-decay`

File: src/bin/asr_train.py [hunk 7]
```diff
@@ -117,8 +131,16 @@ def main():
...
+                        help='Learning rate decay (lr*decay) constant for sgd optimizer')
+    parser.add_argument('--mom', default=0.9, type=float,
```
Comment (sw005320) [inline src/bin/asr_train.py]: `--mom` -> `--momentum` is better to understand the meaning of options.

File: src/bin/asr_train.py [hunk 8]
```diff
@@ -117,8 +131,16 @@ def main():
...
+                        help='Momentum constant for sgd optimizer')
+    parser.add_argument('--wd', default=0, type=float,
```
Comment (sw005320) [inline src/bin/asr_train.py]: better to use more detailed option name than `--wd`

File: src/nets/e2e_asr_attctc_th.py [hunk 9]
```diff
@@ -26,10 +26,15 @@
...
 from e2e_asr_common import label_smoothing_dist
 
+
+torch_is_old = torch.__version__.startswith("0.3.")
+
 CTC_LOSS_THRESHOLD = 10000
 CTC_SCORING_RATIO = 1.5
 MAX_DECODER_OUTPUT = 5
 
+torch_is_old = torch.__version__.startswith("0.3.")
```
Comment (sw005320) [inline src/nets/e2e_asr_attctc_th.py]: Do we need to do the same line twice? (`torch_is_old = torch.__version__.startswith("0.3.")`)

**Comment by sw005320** [comment]
@creatorscan, some of your new functions (e.g., freeze parameters or SGD) can be useful for the other people. I may ask you to make this change as a separate PR to the master branch, later.

**Comment by creatorscan** [comment]
@sw005320 Sure Shinji, Thanks for the response.

**Pull request merged**